File: ios/Sources/CapacitorOidcPlugin/CapacitorOidcPlugin.swift
import AuthenticationServices
import Capacitor
import Foundation

@MainActor
private final class AuthenticationPresentationContextProvider: NSObject, ASWebAuthenticationPresentationContextProviding {
    private let anchor: ASPresentationAnchor

    init(anchor: ASPresentationAnchor) {
        self.anchor = anchor
    }

    func presentationAnchor(for session: ASWebAuthenticationSession) -> ASPresentationAnchor {
        anchor
    }
}

@objc(CapacitorOidcPlugin)
public final class CapacitorOidcPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "CapacitorOidcPlugin"
    public let jsName = "CapacitorOidc"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "configure", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "open", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "cancel", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "storageSet", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "storageGet", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "storageRemove", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "storageGetAllKeys", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setSessionSnapshot", returnType: CAPPluginReturnPromise)
    ]

    private var accessGroup: String?
    private var accessibility = TokenVaultAccessibility.afterFirstUnlockThisDeviceOnly
    @MainActor private var authenticationSession: ASWebAuthenticationSession?
    @MainActor private var authenticationSessionID: UUID?
    @MainActor private var authenticationCall: CAPPluginCall?
    @MainActor private var authenticationPresentationContextProvider: AuthenticationPresentationContextProvider?

    @objc public func configure(_ call: CAPPluginCall) {
        accessGroup = call.getString("keychainAccessGroup")
        if call.getString("keychainAccessibility") == "whenUnlockedThisDeviceOnly" {
            accessibility = .whenUnlockedThisDeviceOnly
        } else {
            accessibility = .afterFirstUnlockThisDeviceOnly
        }
        call.resolve()
    }

    @objc public func open(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            self.startAuthentication(call)
        }
    }

    @objc public func cancel(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            self.cancelAuthentication(call)
        }
    }

    @MainActor
    private func startAuthentication(_ call: CAPPluginCall) {
        guard authenticationSession == nil else {
            call.reject("An authentication session is already active", "AUTH_SESSION_IN_PROGRESS")
            return
        }
        guard let urlValue = call.getString("url"), let url = URL(string: urlValue),
              let callbackValue = call.getString("callbackUrl"), let callbackURL = URL(string: callbackValue),
              let scheme = callbackURL.scheme else {
            call.reject("The authentication URL or callback URL is invalid", "INVALID_CALLBACK")
            return
        }
        guard isSecureRequestURL(url) else {
            call.reject("The authentication URL must use HTTPS outside loopback development", "BROWSER_UNAVAILABLE")
            return
        }
        guard let presentationAnchor = bridge?.viewController?.viewIfLoaded?.window,
              presentationAnchor.windowScene?.activationState == .foregroundActive else {
            call.reject("The app has no active presentation window", "BROWSER_UNAVAILABLE")
            return
        }

        let sessionID = UUID()
        let completion: ASWebAuthenticationSession.CompletionHandler = { [weak self] callbackURL, error in
            DispatchQueue.main.async {
                self?.completeAuthentication(sessionID: sessionID, callbackURL: callbackURL, error: error)
            }
        }

        let session: ASWebAuthenticationSession
        if #available(iOS 17.4, *), scheme.lowercased() == "https" {
            guard let host = callbackURL.host else {
                call.reject("The HTTPS callback URL has no host", "INVALID_CALLBACK")
                return
            }
            session = ASWebAuthenticationSession(
                url: url,
                callback: .https(host: host, path: callbackURL.path),
                completionHandler: completion
            )
        } else {
            guard scheme.lowercased() != "https" else {
                call.reject("HTTPS callbacks require iOS 17.4 or newer", "BROWSER_UNAVAILABLE")
                return
            }
            session = ASWebAuthenticationSession(url: url, callbackURLScheme: scheme, completionHandler: completion)
        }

        let presentationContextProvider = AuthenticationPresentationContextProvider(anchor: presentationAnchor)
        session.presentationContextProvider = presentationContextProvider
        session.prefersEphemeralWebBrowserSession = call.getBool("prefersEphemeralWebBrowserSession") ?? false
        authenticationCall = call
        authenticationSession = session
        authenticationSessionID = sessionID
        authenticationPresentationContextProvider = presentationContextProvider
        guard session.start() else {
            authenticationCall = nil
            authenticationSession = nil
            authenticationSessionID = nil
            authenticationPresentationContextProvider = nil
            call.reject("The system authentication session could not start", "BROWSER_UNAVAILABLE")
            return
        }
    }

    @MainActor
    private func cancelAuthentication(_ call: CAPPluginCall) {
        let activeCall = authenticationCall
        let activeSession = authenticationSession
        authenticationCall = nil
        authenticationSession = nil
        authenticationSessionID = nil
        authenticationPresentationContextProvider = nil
        activeSession?.cancel()
        activeCall?.reject("The user cancelled authentication", "USER_CANCELLED")
        call.resolve()
    }

    @objc public func storageSet(_ call: CAPPluginCall) {
        guard let namespace = call.getString("namespace"), let key = call.getString("key"),
              let value = call.getString("value") else {
            call.reject("Storage input is invalid", "SECURE_STORAGE_ERROR")
            return
        }
        resolveStorage(call) { try vault().set(namespace: namespace, key: key, value: value); return [:] }
    }

    @objc public func storageGet(_ call: CAPPluginCall) {
        guard let namespace = call.getString("namespace"), let key = call.getString("key") else {
            call.reject("Storage input is invalid", "SECURE_STORAGE_ERROR")
            return
        }
        resolveStorage(call) { ["value": try vault().get(namespace: namespace, key: key) as Any] }
    }

    @objc public func storageRemove(_ call: CAPPluginCall) {
        guard let namespace = call.getString("namespace"), let key = call.getString("key") else {
            call.reject("Storage input is invalid", "SECURE_STORAGE_ERROR")
            return
        }
        resolveStorage(call) { ["value": try vault().remove(namespace: namespace, key: key) as Any] }
    }

    @objc public func storageGetAllKeys(_ call: CAPPluginCall) {
        guard let namespace = call.getString("namespace") else {
            call.reject("Storage input is invalid", "SECURE_STORAGE_ERROR")
            return
        }
        resolveStorage(call) { ["keys": try vault().getAllKeys(namespace: namespace)] }
    }

    @objc public func setSessionSnapshot(_ call: CAPPluginCall) {
        guard let namespace = call.getString("namespace") else {
            call.reject("Storage input is invalid", "SECURE_STORAGE_ERROR")
            return
        }
        resolveStorage(call) {
            try vault().setSessionSnapshot(namespace: namespace, value: call.getString("value"))
            return [:]
        }
    }

    @MainActor
    private func completeAuthentication(sessionID: UUID, callbackURL: URL?, error: Error?) {
        guard authenticationSessionID == sessionID else { return }
        let call = authenticationCall
        authenticationCall = nil
        authenticationSession = nil
        authenticationSessionID = nil
        authenticationPresentationContextProvider = nil

        if let sessionError = error as? ASWebAuthenticationSessionError, sessionError.code == .canceledLogin {
            call?.reject("The user cancelled authentication", "USER_CANCELLED")
        } else if error != nil {
            call?.reject("The system authentication session failed", "BROWSER_UNAVAILABLE")
        } else if let callbackURL {
            call?.resolve(["url": callbackURL.absoluteString])
        } else {
            call?.reject("The authentication session returned no callback", "INVALID_CALLBACK")
        }
    }

    private func vault() -> TokenVault {
        TokenVault(accessGroup: accessGroup, accessibility: accessibility)
    }

    private func isSecureRequestURL(_ url: URL) -> Bool {
        if url.scheme?.lowercased() == "https" { return true }
        guard url.scheme?.lowercased() == "http", let host = url.host?.lowercased() else { return false }
        return host == "localhost" || host == "127.0.0.1" || host == "::1"
    }

    private func resolveStorage(_ call: CAPPluginCall, operation: () throws -> [String: Any]) {
        do {
            call.resolve(try operation())
        } catch {
            call.reject("Secure storage failed", "SECURE_STORAGE_ERROR")
        }
    }
}

File: android/src/main/java/com/jojopirker/capacitor/oidc/CapacitorOidcPlugin.java
package com.jojopirker.capacitor.oidc;

import android.content.ActivityNotFoundException;
import android.content.Intent;
import android.net.Uri;
import androidx.activity.result.ActivityResultLauncher;
import androidx.browser.auth.AuthTabIntent;
import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import java.io.IOException;
import java.security.GeneralSecurityException;
import java.util.List;
import org.json.JSONObject;

@CapacitorPlugin(name = "CapacitorOidc")
public final class CapacitorOidcPlugin extends Plugin {

    private static final String AUTH_SESSION_IN_PROGRESS = "AUTH_SESSION_IN_PROGRESS";
    private static final String USER_CANCELLED = "USER_CANCELLED";
    private static final String BROWSER_UNAVAILABLE = "BROWSER_UNAVAILABLE";
    private static final String INVALID_CALLBACK = "INVALID_CALLBACK";
    private static final String SECURE_STORAGE_ERROR = "SECURE_STORAGE_ERROR";

    private ActivityResultLauncher<Intent> authLauncher;
    private PluginCall pendingAuthCall;
    private Uri pendingCallback;
    private TokenVault vault;

    @Override
    public void load() {
        vault = new TokenVault(getContext());
        authLauncher = bridge.registerForActivityResult(new AuthTabIntent.AuthenticateUserResultContract(), this::handleAuthResult);
    }

    @PluginMethod
    public void configure(PluginCall call) {
        call.resolve();
    }

    @PluginMethod
    public void open(PluginCall call) {
        if (pendingAuthCall != null) {
            call.reject("An authentication session is already active.", AUTH_SESSION_IN_PROGRESS);
            return;
        }

        String urlValue = requiredString(call, "url");
        String callbackValue = requiredString(call, "callbackUrl");
        if (urlValue == null || callbackValue == null) return;

        Uri url = Uri.parse(urlValue);
        Uri callback = Uri.parse(callbackValue);
        if (!isSecureRequestUrl(url)) {
            call.reject("The authentication URL must use HTTPS outside loopback development.", BROWSER_UNAVAILABLE);
            return;
        }
        if (!isSupportedCallback(callback)) {
            call.reject("The callback URL must use HTTPS or a custom scheme.", INVALID_CALLBACK);
            return;
        }

        boolean ephemeral = Boolean.TRUE.equals(call.getBoolean("prefersEphemeralWebBrowserSession", false));
        AuthTabIntent authTab = new AuthTabIntent.Builder().setEphemeralBrowsingEnabled(ephemeral).build();
        pendingAuthCall = call;
        pendingCallback = callback;

        try {
            if ("https".equalsIgnoreCase(callback.getScheme())) {
                authTab.launch(authLauncher, url, callback.getHost(), path(callback));
            } else {
                authTab.launch(authLauncher, url, callback.getScheme());
            }
        } catch (ActivityNotFoundException | IllegalStateException error) {
            clearPendingAuth();
            call.reject("No compatible system browser is available.", BROWSER_UNAVAILABLE);
        }
    }

    @PluginMethod
    public void cancel(PluginCall call) {
        PluginCall authCall = pendingAuthCall;
        clearPendingAuth();
        if (authCall != null) authCall.reject("The authentication session was cancelled.", USER_CANCELLED);
        call.resolve();
    }

    @Override
    protected void handleOnNewIntent(Intent intent) {
        Uri callback = intent.getData();
        PluginCall call = pendingAuthCall;
        if (call == null || callback == null || !CallbackUriMatcher.matches(callback, pendingCallback)) return;

        clearPendingAuth();
        JSObject response = new JSObject();
        response.put("url", callback.toString());
        call.resolve(response);
    }

    @PluginMethod
    public void storageSet(PluginCall call) {
        String namespace = requiredString(call, "namespace");
        String key = requiredString(call, "key");
        String value = call.getString("value");
        if (namespace == null || key == null) return;
        if (value == null) {
            call.reject("value is required.");
            return;
        }

        try {
            vault.set(namespace, key, value);
            call.resolve();
        } catch (GeneralSecurityException | IOException error) {
            rejectStorage(call);
        }
    }

    @PluginMethod
    public void storageGet(PluginCall call) {
        String namespace = requiredString(call, "namespace");
        String key = requiredString(call, "key");
        if (namespace == null || key == null) return;

        try {
            resolveValue(call, vault.get(namespace, key));
        } catch (GeneralSecurityException | IOException error) {
            rejectStorage(call);
        }
    }

    @PluginMethod
    public void storageRemove(PluginCall call) {
        String namespace = requiredString(call, "namespace");
        String key = requiredString(call, "key");
        if (namespace == null || key == null) return;

        try {
            resolveValue(call, vault.remove(namespace, key));
        } catch (GeneralSecurityException | IOException error) {
            rejectStorage(call);
        }
    }

    @PluginMethod
    public void storageGetAllKeys(PluginCall call) {
        String namespace = requiredString(call, "namespace");
        if (namespace == null) return;

        try {
            List<String> keys = vault.getAllKeys(namespace);
            JSObject result = new JSObject();
            result.put("keys", new JSArray(keys));
            call.resolve(result);
        } catch (GeneralSecurityException | IOException error) {
            rejectStorage(call);
        }
    }

    @PluginMethod
    public void setSessionSnapshot(PluginCall call) {
        String namespace = requiredString(call, "namespace");
        if (namespace == null) return;

        try {
            vault.setSessionSnapshot(namespace, call.getString("value"));
            call.resolve();
        } catch (GeneralSecurityException | IOException error) {
            rejectStorage(call);
        }
    }

    private void handleAuthResult(AuthTabIntent.AuthResult result) {
        PluginCall call = pendingAuthCall;
        Uri expectedCallback = pendingCallback;
        clearPendingAuth();
        if (call == null) return;

        if (result.resultCode == AuthTabIntent.RESULT_CANCELED) {
            call.reject("The authentication session was cancelled.", USER_CANCELLED);
            return;
        }
        if (result.resultCode != AuthTabIntent.RESULT_OK || result.resultUri == null) {
            call.reject("The browser did not return a valid callback.", INVALID_CALLBACK);
            return;
        }
        if (!CallbackUriMatcher.matches(result.resultUri, expectedCallback)) {
            call.reject("The authentication callback does not match the configured redirect URL.", INVALID_CALLBACK);
            return;
        }

        JSObject response = new JSObject();
        response.put("url", result.resultUri.toString());
        call.resolve(response);
    }

    private void clearPendingAuth() {
        pendingAuthCall = null;
        pendingCallback = null;
    }

    private static String requiredString(PluginCall call, String name) {
        String value = call.getString(name);
        if (value == null || value.isEmpty()) {
            call.reject(name + " is required.");
            return null;
        }
        return value;
    }

    private static boolean isSecureRequestUrl(Uri uri) {
        String scheme = uri.getScheme();
        String host = uri.getHost();
        if (host == null) return false;
        if ("https".equalsIgnoreCase(scheme)) return true;
        return "http".equalsIgnoreCase(scheme) &&
        ("localhost".equalsIgnoreCase(host) || "127.0.0.1".equals(host) || "::1".equals(host));
    }

    private static boolean isSupportedCallback(Uri uri) {
        String scheme = uri.getScheme();
        if (scheme == null || scheme.isEmpty() || "http".equalsIgnoreCase(scheme)) return false;
        return !"https".equalsIgnoreCase(scheme) || uri.getHost() != null;
    }

    private static String path(Uri uri) {
        String path = uri.getPath();
        return path == null ? "" : path;
    }

    private static void resolveValue(PluginCall call, String value) {
        JSObject result = new JSObject();
        result.put("value", value == null ? JSONObject.NULL : value);
        call.resolve(result);
    }

    private static void rejectStorage(PluginCall call) {
        call.reject("Secure storage failed.", SECURE_STORAGE_ERROR);
    }
}
